runtime/lua/vim/diagnostic.lua

```diff
@@ -1857,16 +1857,37 @@ function M.open_float(opts, ...)
   if scope == 'line' then
     --- @param d vim.Diagnostic
     diagnostics = vim.tbl_filter(function(d)
-      return lnum >= d.lnum and lnum <= d.end_lnum
+      if lnum < d.lnum or lnum > d.end_lnum then
+        return false
+      end
+      if d.lnum == d.end_lnum then
+        -- if diagnostic only spans one line, always show
+        -- allows showing diagnostics with zero width (where lnum == end_lnum and col == end_col)
+        return true
+      else
+        -- end-exclusive range, don't show if on end_lnum and end_col == 0
+        return lnum ~= d.end_lnum or d.end_col ~= 0
+      end
     end, diagnostics)
   elseif scope == 'cursor' then
-    -- LSP servers can send diagnostics with `end_col` past the length of the line
-    local line_length = #api.nvim_buf_get_lines(bufnr, lnum, lnum + 1, true)[1]
+    -- LSP servers can send diagnostics with `col` past the end of the line
+    local lnum_length = #api.nvim_buf_get_lines(bufnr, lnum, lnum + 1, true)[1]
```

Comment:
Any reason to rename this variable? The old name `line_length` was fine, the new name `lnum_length` is more confusing (`lnum` is the line number, so `lnum_length` doesn't really make sense).

Reply:
The reasoning behind that was that the length is for the line referred to by `d.lnum`, if the diagnostic was multiple lines long. But now that I think about it, that is more confusing - I'll revert it back.